### PR TITLE
Alter the way HDCED4+ is calculated for CRS-1787

### PR DESCRIPTION
**Pull request by jbowdidge-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/ConsecutiveSentence.kt [hunk 1]
```diff
@@ -188,6 +188,10 @@ class ConsecutiveSentence(val orderedSentences: List<CalculableSentence>) : Calc
   }
 
   fun isMadeUpOfSdsAndSdsPlusSentences(): Boolean {
-    return (orderedSentences.any { it.offence.isPcscSdsPlus } || orderedSentences.any { it.offence.isPcscSds }) && orderedSentences.any { !it.offence.isPcscSdsPlus }
+    return orderedSentences.any { it.isSdsPlus() } && orderedSentences.any { !it.isSdsPlus() }
+  }
+
+  override fun isSdsPlus(): Boolean {
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/ConsecutiveSentence.kt:194]: isMadeUpOnlySdsPlus?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SentenceIdentificationService.kt [hunk 2]
```diff
@@ -345,7 +345,7 @@ class SentenceIdentificationService(val hdcedCalculator: HdcedCalculator, val tu
             }
           }
         } else {
-          if (overEighteen && durationGreaterThanSevenYears && sentence.offence.isScheduleFifteenMaximumLife) {
+          if (sentence.sentencedAt.isAfterOrEqualTo(SDS_PLUS_COMMENCEMENT_DATE) && overEighteen && durationGreaterThanSevenYears && sentence.offence.isScheduleFifteenMaximumLife) {
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SentenceIdentificationService.kt:348]: This should probably be inside the other if block?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 3]
```diff
@@ -49,9 +49,9 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
       .minus(sentenceCalculation.calculatedUnusedReleaseAda)
 
     // Now that we have sentences > 4 years there are rules around how Consecutive sentences containing SDS+ are calculated
-    if (sentence is ConsecutiveSentence && sentence.isMadeUpOfSdsAndSdsPlusSentences() && nonSdsPlusGreaterThanMinimum(sentence) && nonSdsLessThanMidpoint(sentence) && !lastSentenceIsSdsPlus(sentence)) {
-      val latestNcrd = sentence.orderedSentences.filter { it.offence.isPcscSdsPlus || it.offence.isPcscSds }.maxBy { it.sentenceCalculation.releaseDate }
-      val nonSdsPlusSentences = sentence.orderedSentences.filter { !it.offence.isPcscSdsPlus || !it.offence.isPcscSds }
+    if (sentence is ConsecutiveSentence && sentence.isSdsPlus() && nonSdsPlusGreaterThanMinimum(sentence) && nonSdsLessThanMidpoint(sentence) && !lastSentenceIsSdsPlus(sentence)) {
+      val latestNcrd = sentence.orderedSentences.filter { it.isSdsPlus() }.maxBy { it.sentenceCalculation.releaseDate }
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:53]: Notional CRD?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 4]
```diff
@@ -49,9 +49,9 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
       .minus(sentenceCalculation.calculatedUnusedReleaseAda)
 
     // Now that we have sentences > 4 years there are rules around how Consecutive sentences containing SDS+ are calculated
-    if (sentence is ConsecutiveSentence && sentence.isMadeUpOfSdsAndSdsPlusSentences() && nonSdsPlusGreaterThanMinimum(sentence) && nonSdsLessThanMidpoint(sentence) && !lastSentenceIsSdsPlus(sentence)) {
-      val latestNcrd = sentence.orderedSentences.filter { it.offence.isPcscSdsPlus || it.offence.isPcscSds }.maxBy { it.sentenceCalculation.releaseDate }
-      val nonSdsPlusSentences = sentence.orderedSentences.filter { !it.offence.isPcscSdsPlus || !it.offence.isPcscSds }
+    if (sentence is ConsecutiveSentence && sentence.isSdsPlus() && nonSdsPlusGreaterThanMinimum(sentence) && nonSdsLessThanMidpoint(sentence) && !lastSentenceIsSdsPlus(sentence)) {
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:52]: Haven't you negated doing the rest of this if by checking that all the sentences are SDS+?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 5]
```diff
@@ -62,23 +62,45 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
...
+        }
+      } else {
+        val nonSdsPluSentences = sentence.orderedSentences.filter { !it.isSdsPlus() }
+        val nonSdsPlusSentenceLengthInDays = nonSdsPluSentences.sumOf { it.getLengthInDays() }.toLong()
+
+        if (nonSdsPlusSentenceLengthInDays < hdcedConfiguration.envelopeMidPoint) {
+          val quarterSentenceLengthOr28Days = if (ceil(sentenceCalculation.numberOfDaysToSentenceExpiryDate / 4.0).toLong() > 28) ceil(sentenceCalculation.numberOfDaysToSentenceExpiryDate / 4.0).toLong() else 28
+          log.info("Quarter sentence length: {}", quarterSentenceLengthOr28Days)
+          sentenceCalculation.numberOfDaysToHomeDetentionCurfew4PlusEligibilityDate = quarterSentenceLengthOr28Days.plus(adjustedDays)
+          sentenceCalculation.homeDetentionCurfew4PlusEligibilityDate =
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:96]: isn't this a pointless assignment?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 6]
```diff
@@ -62,23 +62,45 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
...
+            sentence.sentencedAt.plusDays(sentenceCalculation.numberOfDaysToHomeDetentionCurfew4PlusEligibilityDate)
+          sentenceCalculation.homeDetentionCurfew4PlusEligibilityDate = sentenceCalculation.adjustedDeterminateReleaseDate.plusDays(quarterSentenceLengthOr28Days)
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:98]: Shouldn't this be sentence date +  1/4 sentence?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 7]
```diff
@@ -62,23 +62,45 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
...
+          sentenceCalculation.homeDetentionCurfew4PlusEligibilityDate = sdsPlusNotionalCrd.plusDays(quarterSentenceLengthOr28Days).minusDays(sentence.sentenceCalculation.calculatedTotalDeductedDays.toLong()).plusDays(sentence.sentenceCalculation.calculatedTotalAddedDays.toLong())
+        } else {
+          sentenceCalculation.homeDetentionCurfew4PlusEligibilityDate = sdsPlusNotionalCrd.minusDays(hdcedConfiguration.deductionDays)
+        }
+      } else {
+        val nonSdsPluSentences = sentence.orderedSentences.filter { !it.isSdsPlus() }
+        val nonSdsPlusSentenceLengthInDays = nonSdsPluSentences.sumOf { it.getLengthInDays() }.toLong()
+
+        if (nonSdsPlusSentenceLengthInDays < hdcedConfiguration.envelopeMidPoint) {
+          val quarterSentenceLengthOr28Days = if (ceil(sentenceCalculation.numberOfDaysToSentenceExpiryDate / 4.0).toLong() > 28) ceil(sentenceCalculation.numberOfDaysToSentenceExpiryDate / 4.0).toLong() else 28
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:93]: Done differently here to further down in  `calculateHdcedUnderMidpoint`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 8]
```diff
@@ -32,7 +32,7 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
 
   fun doesHdced4DateApply(sentence: CalculableSentence, offender: Offender, isMadeUpOfOnlyDtos: Boolean): Boolean {
     return sentence.durationIsGreaterThanOrEqualTo(hdcedConfiguration.envelopeMinimum, hdcedConfiguration.envelopeMinimumUnit) &&
-      !offender.isActiveSexOffender && !isMadeUpOfOnlyDtos && ((sentence !is ConsecutiveSentence && !sentence.offence.isPcscSdsPlus) || (sentence is ConsecutiveSentence && sentence.isMadeUpOfSdsAndSdsPlusSentences()))
+      !offender.isActiveSexOffender && !isMadeUpOfOnlyDtos && !sentence.isSdsPlus()
```
[the remark this answers is not in the document]
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:35]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 9]
```diff
@@ -62,23 +62,45 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
...
-      val sdsNotionalSled = sdsPlusNotionalCrd.plusDays(lengthInDaysOfNonSdsPlusSentences)
-      log.info("Sds notional sled: {}", sdsNotionalSled)
-      val quarterSentenceLength = ceil(lengthInDaysOfNonSdsPlusSentences / 4.0).toLong()
-      log.info("Quarter sentence length: {}", quarterSentenceLength)
-
-      sentenceCalculation.homeDetentionCurfew4PlusEligibilityDate = sdsPlusNotionalCrd.plusDays(quarterSentenceLength).minusDays(sentence.sentenceCalculation.calculatedTotalDeductedDays.toLong())
+    } else if (sentence is ConsecutiveSentence && !sentence.isSdsPlus()) {
+      if (sentence.orderedSentences.any { it.isSdsPlus() }) {
+        val sdsPlusSentences = sentence.orderedSentences.filter { it.isSdsPlus() }
+        val nonSdsPluSentences = sentence.orderedSentences.filter { !it.isSdsPlus() }
```
[the remark this answers is not in the document]
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:68]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt [hunk 10]
```diff
@@ -62,23 +62,45 @@ class Hdced4Calculator(val hdcedConfiguration: Hdced4Configuration) {
...
+        log.info("Sds plus notional sled: {}", sdsPlusNotionalSled)
+        val sdsPlusNotionalCrd = sentence.sentencedAt.minusDays(1).plusDays(ceil(sdsPlusLengthInDays * (2.0 / 3.0)).toLong())
+        log.info("Sds plus notional crd: {}", sdsPlusNotionalCrd)
+        val lengthInDaysOfNonSdsPlusSentences = sentence.orderedSentences.filter { !it.isSdsPlus() }.sumOf { it.totalDuration().getLengthInDays(sdsPlusNotionalCrd) }.toLong()
+        log.info("Length in days of non sds plus sentences: {}", lengthInDaysOfNonSdsPlusSentences)
+
+        val sdsNotionalSled = sdsPlusNotionalCrd.plusDays(lengthInDaysOfNonSdsPlusSentences)
+        log.info("Sds notional sled: {}", sdsNotionalSled)
+        if (nonSdsPlusSentenceLengthInDays < hdcedConfiguration.envelopeMidPoint) {
+          val quarterSentenceLengthOr28Days = if (ceil(lengthInDaysOfNonSdsPlusSentences / 4.0).toLong() > 28) ceil(lengthInDaysOfNonSdsPlusSentences / 4.0).toLong() else 28
```
[the remark this answers is not in the document]
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:82]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/ConsecutiveSentence.kt (hunk 1 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/ConsecutiveSentence.kt:194]: Leaving as discussed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt (hunk 3 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:53]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SentenceIdentificationService.kt (hunk 2 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SentenceIdentificationService.kt:348]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt (hunk 6 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:98]: Hmm

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt (hunk 4 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:52]: Yes - I was doing some playing around with the code, and forgot to remove this. Great spot.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt [hunk 11]
```diff
@@ -515,7 +515,7 @@ data class SentenceCalculation(
       "PED"
     }
 
-    return "Date of $expiryDateType\t:\t${unadjustedExpiryDate.format(formatter)}\n" +
+    return "Sentence type: ${sentence.javaClass.name}\n" + "Date of $expiryDateType\t:\t${unadjustedExpiryDate.format(formatter)}\n" +
```
[the remark this answers is not in the document]
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:518]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt (hunk 6 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt:98]: Great spot - thank you for questioning this!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt (hunk 7 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/Hdced4Calculator.kt]: Done.

**Review by srowlands-moj — APPROVED** [review]
LGTM

**Comment by srowlands-moj** [comment]
LGTM - a couple of minor typos we can amend but don't want to delay testing further

**Review by srowlands-moj — APPROVED** [review]

**Pull request merged**